### DISCUSSION TITLE
Performance about SilkTouch

### DISCUSSION THREAD

**Discussion by BobbyBao** [dotnet/Silk.NET#448] [Q&A] [answered]
This is the decompiled code, will this cause performance loss?
	
	[NativeApi(EntryPoint = "vkCmdDrawIndexed")]
	public unsafe void CmdDrawIndexed([Count(Count = 0)] CommandBuffer commandBuffer, [Count(Count = 0)] uint indexCount, [Count(Count = 0)] uint instanceCount, [Count(Count = 0)] uint firstIndex, [Count(Count = 0)] int vertexOffset, [Count(Count = 0)] uint firstInstance)
	{
		((delegate* unmanaged[Cdecl]<CommandBuffer, uint, uint, uint, int, uint, void>)(void*)(base.CurrentVTable as GeneratedVTable).Load(189, "vkCmdDrawIndexed"))(commandBuffer, indexCount, instanceCount, firstIndex, vertexOffset, firstInstance);
	}

**Comment by HurricanKai** [comment]
What about this? We generate the method bodies and they are highly optimized. If you have questions, I'm happy to answer them.

**Comment by BobbyBao** [comment]
// {
		IL_0000: nop
		// (no C# code)
		IL_0001: nop
		// ((delegate* unmanaged[Cdecl]<CommandBuffer, uint, uint, uint, int, uint, void>)(void*)(base.CurrentVTable as GeneratedVTable).Load(189, "vkCmdDrawIndexed"))(commandBuffer, indexCount, instanceCount, firstIndex, vertexOffset, firstInstance);
		IL_0002: nop
		IL_0003: ldarg.0
		IL_0004: call instance class [Silk.NET.Core]Silk.NET.Core.Native.IVTable [Silk.NET.Core]Silk.NET.Core.Native.NativeApiContainer::get_CurrentVTable()
		IL_0009: isinst Silk.NET.Vulkan.Vk/GeneratedVTable
		IL_000e: ldc.i4 189
		IL_0013: ldstr "vkCmdDrawIndexed"
		IL_0018: callvirt instance native int Silk.NET.Vulkan.Vk/GeneratedVTable::Load(int32, string)
		IL_001d: call void* [System.Runtime]System.IntPtr::op_Explicit(native int)
		IL_0022: stloc.0
		IL_0023: ldarg.1
		IL_0024: ldarg.2
		IL_0025: ldarg.3
		IL_0026: ldarg.s firstIndex
		IL_0028: ldarg.s vertexOffset
		IL_002a: ldarg.s firstInstance
		IL_002c: ldloc.0
		// (no C# code)
		IL_002d: calli unmanaged cdecl void(valuetype Silk.NET.Vulkan.CommandBuffer, uint32, uint32, uint32, int32, uint32)
		IL_0032: nop
		// }
		IL_0033: nop
		IL_0034: ret
		
		These are  unnecessary instructions:
		
		IL_0003: ldarg.0
		IL_0004: call instance class [Silk.NET.Core]Silk.NET.Core.Native.IVTable [Silk.NET.Core]Silk.NET.Core.Native.NativeApiContainer::get_CurrentVTable()
		IL_0009: isinst Silk.NET.Vulkan.Vk/GeneratedVTable
		IL_000e: ldc.i4 189
		IL_0013: ldstr "vkCmdDrawIndexed"
		IL_0018: callvirt instance native int Silk.NET.Vulkan.Vk/GeneratedVTable::Load(int32, string)
		IL_001d: call void* [System.Runtime]System.IntPtr::op_Explicit(native int)

**Comment by HurricanKai** [comment, marked as answer]
you are inspecting IL here - while our bindings aren't the most IL space efficient, the JIT is very good at turning these into assembly, and they are chosen to allow JIT to optimize. 
You are referencing the seemingly uncessary cast to Silk.NET.Vulkan.Vk.GeneratedVTable, but this is done to allow the JIT to assume the type and devirtualize the call, allowing inlining of the call
in asm, a typical function will look something like
```
G_M60490_IG05:
   mov      rax, qword ptr [rsi+16]
   mov      rcx, rax
   test     rcx, rcx
   je       SHORT G_M60490_IG07
G_M60490_IG06:
   jmp      SHORT G_M60490_IG08
G_M60490_IG07:
   mov      rcx, gword ptr [rsi+8]
   mov      rdx, 0xD1FFAB1E
   mov      rdx, gword ptr [rdx]
   mov      r11, 0xD1FFAB1E
   call     [Silk.NET.Core.Contexts.INativeContext:GetProcAddress(System.String):long:this]
   mov      qword ptr [rsi+16], rax
   mov      rax, qword ptr [rsi+16]
   
G_M60490_IG08:
   mov      rcx, rdi
   mov      rdx, rbx
   mov      r8, r14
   mov      r9d, dword ptr [rbp+30H]
   lea      r10, G_M60490_IG10
   mov      qword ptr [rbp-50H], r10
   lea      r10, bword ptr [rbp-78H]
   mov      qword ptr [r15+16], r10
   mov      byte  ptr [r15+12], 0
G_M60490_IG09:
   call     rax
```
and usually the `G_M60490_IG07` part will be skipped (it's the "lazy loading" part, so usually only executed once)
(Note: this ASM is a bit older, I've taken it from a file I had laying around when I optimized this (.NET 5 preview ?), if you want to, I can get a newer one for .NET 6, but I don't expect there to be any differences)

you also mention the `nint` to `void*` cast in that IL, this won't influence JIT asm at all, since that is just a cast to allow type safety, at the assembly level that can be "ignored"
(`void*` is a pointer, and `nint` is a pointer sized int, so there can be no overhead casting those)
